master/buildbot/www/hooks/bitbucketserver.py

```diff
@@ -0,0 +1,168 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+# Copyright Mamba Team
+import logging
+from dateutil.parser import parse as dateparse
+
+from twisted.python import log
+
+import json
+
+GIT_MERGE_REF = "refs/pull-requests/{}/merge"
+GIT_HEAD_REF = "refs/heads/{}"
+
+_HEADER_CT = 'Content-Type'
+_HEADER_EVENT = 'X-Event-Key'
+
+class BitbucketServerEventHandler(object):
+
+    def __init__(self, codebase=None, options={}):
+        self._codebase = codebase
+        self.options = options
+
+    def process(self, request):
+        payload = self._get_payload(request)
+        event_type = request.getHeader(_HEADER_EVENT)
+        log.msg("Processing event {header}: {event}"
+                .format(header=_HEADER_EVENT, event=event_type))
+        event_type = event_type.replace(":", "_")
+        handler = getattr(self, 'handle_%s' % event_type, None)
+
+        if handler is None:
+            raise ValueError('Unknown event: {}'.format(event_type))
+
+        return handler(payload)
+
+    def _get_payload(self, request):
+        content = request.content.read()
+        content_type = request.getHeader(_HEADER_CT)
+        if content_type.startswith('application/json'):
+            payload = json.loads(content)
+        elif content_type.startswith('application/x-www-form-urlencoded'):
+            payload = json.loads(request.args['payload'][0])
+        else:
+            raise ValueError('Unknown content type: {}'
+                             .format(content_type))
+
+        log.msg("Payload: {}".format(payload))
+
+        return payload
+
+    def handle_repo_push(self, payload):
+        changes = []
+        project = payload['repository']['project']['name']
+        repo_url = payload['repository']['links']['self'][0]['href']
+        repo_url = repo_url.rstrip('browse')
+
+        for payload_change in payload['push']['changes']:
+            if not payload_change['new']:
+                # skip change if deleting a tag or a branch
+                continue
+
+            change = {
```

Comment:
You've switched to `%s` formatting in this dictionary. `handle_pullrequest` constructs a similar dictionary using `{}`